### PR TITLE
librbd/cache/pwl/ssd: make log entry 64 bit and add ssd version control

### PR DESCRIPTION
**Pull request by CongMinYin**
in src/librbd/cache/pwl/ssd/WriteLog.cc:load_existing_entries()
```
    uint64_t next_log_pos = this->m_first_valid_entry;
    uint64_t curr_log_pos = next_log_pos; 
    log_entry->log_entry_index = curr_log_pos;  
```
so change `log_entry_index` to `uint64_t`

in src/librbd/cache/pwl/ssd/WriteLog.cc:initialize_pool()
```
    new_root->first_free_entry = m_first_free_entry;
    new_root->first_valid_entry = m_first_valid_entry;
```
so change `first_free_entry` and `first_valid_entry` to `uint64_t`

in src/librbd/cache/pwl/ssd/WriteLog.cc:retire_entries()
```
    initial_first_valid_entry = m_first_valid_entry;
    first_valid_entry = m_first_valid_entry;
```
 so change `initial_first_valid_entry` and `first_valid_entry` to `uint64_t`

Fixes: https://tracker.ceph.com/issues/50675

Add ssd version control.

Signed-off-by: Yin Congmin <congmin.yin@intel.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/librbd/cache/pwl/AbstractWriteLog.cc
Comment (MahatiC) [inline src/librbd/cache/pwl/AbstractWriteLog.cc:1467]: these values are `rwl` specific, so they can be 32bit since the log entries are limited to a million

**Review by MahatiC — COMMENTED** [review]
I think the `uint64_t` change is only relevant to `ssd` mode, not `rwl`

File: src/librbd/cache/pwl/AbstractWriteLog.cc
Comment (CongMinYin) [reply src/librbd/cache/pwl/AbstractWriteLog.cc:1467]: > these values are `rwl` specific, so they can be 32bit since the log entries are limited to a million

The types here are so confusing that I think it's necessary to make an alignment. Its statement is `uint64_t`, like `uint64_t num_lanes = 0;`. The parameter type passed func is `uint32_t`, like `uint32_t num_lanes`, but the variables involved in the operation are `uint64_t` and `unsigned int`, like 
```
      m_free_lanes -= num_lanes;
      m_free_log_entries -= num_log_entries;
```
`m_free_lanes` is `unsigned int`.
 `m_free_log_entries` is `uint32_t`, it must be change to `uint64_t`.

Except some temporary variables, I think we should not use the compiler dependent types like `unsign int`, `int`, `long` and so on. change these to `uint64_t` and `uint32_t`.
@idryomov What's your opinion? ? I'll make a further revision according to your opinion or leave it as it is.

Comment (idryomov) [reply src/librbd/cache/pwl/AbstractWriteLog.cc:1467]: `num_lanes`, `num_log_entries` and `num_unpublished_reserves` parameters are compared with `m_free_lanes`, `m_free_log_entries` and `m_unpublished_reserves` all of which are specific to `rwl` mode and are either `uint32_t` or `unsigned int`.  

This PR is supposed to address issues with >4G cache size in `ssd` mode.  Converting some fields from `unsigned int` to `uint32_t` is not a bad idea but should be done separately.  Let's keep this PR as small as possible so that it is clear what exactly was leading to data corruption at 4G boundary.

Comment (idryomov) [reply src/librbd/cache/pwl/AbstractWriteLog.cc:1467]: >  `m_free_log_entries` is `uint32_t`, it must be change to `uint64_t`.

No, `m_free_log_entries` is fine as `uint32_t`.  It is always zero in `ssd` mode and per @MahatiC is limited to a million in `rwl` mode.

The way how some of log entry accounting is shared through `AbstractWriteLog` even though log entries in `ssd` mode are very different from log entries in `rwl` mode is super confusing but again it wasn't me who designed this.

**Review by idryomov — DISMISSED** [review]
Please drop changes that don't affect `ssd` mode (`num_log_entries`, `referring_map_entries`, etc).  Let's make this patch as precise as possible highlighting exactly what was leading to data corruption at 4G boundary in `ssd` mode.

File: src/librbd/cache/pwl/Types.h
Comment (idryomov) [inline src/librbd/cache/pwl/Types.h:299]: This is an on-disk format change.  I think we need to bump the layout version for `ssd` mode.

**Comment by CongMinYin** [comment]
> Please drop changes that don't affect `ssd` mode (`num_log_entries`, `referring_map_entries`, etc). Let's make this patch as precise as possible highlighting exactly what was leading to data corruption at 4G boundary in `ssd` mode.

Get to your point, only make necessary changes for this bug, and leave other optimizations for later. 
So the title is a little bit inappropriate. Now it's just to modify `m_first_valid_entry` and `m_first_free_entry` related variables to 64 bit.

File: src/librbd/cache/pwl/Types.h
Comment (idryomov) [reply src/librbd/cache/pwl/Types.h:299]: Hrm, speaking of layout version, it's not even checked `ssd` mode.  Need to add that!

**Comment by idryomov** [comment]
> So the title is a little bit inappropriate. Now it's just to modify `m_first_valid_entry` and `m_first_free_entry` related variables to 64 bit.

Yes, I already changed the title of the PR.  Please update the commit message accordingly.

File: src/librbd/cache/pwl/Types.h
Comment (CongMinYin) [reply src/librbd/cache/pwl/Types.h:299]: > layout

I try to find `layout_version` in ssd mode. but, the author didn't use version control. cc @MahatiC

Comment (CongMinYin) [reply src/librbd/cache/pwl/Types.h:299]: > This is an on-disk format change. I think we need to bump the layout version for `ssd` mode.

I try to find layout_version in ssd mode. but, the author didn't use version control. cc @MahatiC

**Comment by CongMinYin** [comment]
> > So the title is a little bit inappropriate. Now it's just to modify `m_first_valid_entry` and `m_first_free_entry` related variables to 64 bit.
> 
> Yes, I already changed the title of the PR. Please update the commit message accordingly.

done

File: src/librbd/cache/pwl/Types.h
Comment (MahatiC) [reply src/librbd/cache/pwl/Types.h:299]: Yes, layout_version isn't checked in ssd mode, but I wasn't involved in that part of design decision - so I'm not sure why. Will try to dig into docs for any info on this

Comment (CongMinYin) [reply src/librbd/cache/pwl/Types.h:299]: > Hrm, speaking of layout version, it's not even checked `ssd` mode. Need to add that!

Add layout version in this pr, or open a new PR to add in the future?

Comment (idryomov) [reply src/librbd/cache/pwl/Types.h:299]: In this PR -- we can't change the types without bumping the version and we can't bump the version without adding the check.

Comment (CongMinYin) [reply src/librbd/cache/pwl/Types.h:299]: > In this PR -- we can't change the types without bumping the version and we can't bump the version without adding the check. (Of course, the check should have been there from the start...)

I will open a separate PR for layout version. After version control pr merged,  change version here. Is this ok?

Comment (idryomov) [reply src/librbd/cache/pwl/Types.h:299]: I would prefer to do it here.  Currently the version is not only checked but also not set, so we can assume that version 0 stands for "broken".  The 64-bit log entry pointers would be version 1.  No need to create a separate PR for a few lines of code that are very closely related.

**Comment by CongMinYin** [comment]
@idryomov layout version added.

**Comment by CongMinYin** [comment]
jenkins test make check

File: src/librbd/cache/pwl/ssd/WriteLog.cc
Comment (idryomov) [inline src/librbd/cache/pwl/ssd/WriteLog.cc:173]: I know you just moved this code, but need to check for errors here:
```suggestion
    r = bdev->read(0, MIN_WRITE_ALLOC_SSD_SIZE, &bl, &ioctx, false);
    if (r < 0) {
      ...
    }
```

**Review by idryomov — COMMENTED** [review]
Additional concerns:
- `WriteLogCacheEntry::entry_index` is also on-disk and used by the `ssd` mode to store the log entry pointer but is still uint32_t
- `entry_index` parameter of `AbstractWriteLog::update_entry()` is int but the `ssd` mode passes uint64_t

Please perform a thorough review of all fields, parameters and local variables across the entire `src/librbd/cache/pwl` code base to make sure there is nothing else of this sort lingering out there.  Especially the on-disk fields -- I'd rather not change the on-disk format again...

**Comment by idryomov** [comment]
Also, `WriteLogPoolRoot::cur_sync_gen` is stored on disk but doesn't appear to be used anywhere.

@MahatiC what is it for?  Do you intend to use it in the future or should we remove it since we are changing the on-disk format here anyway?

**Comment by MahatiC** [comment]
> Also, `WriteLogPoolRoot::cur_sync_gen` is stored on disk but doesn't appear to be used anywhere.
> 
> @MahatiC what is it for? Do you intend to use it in the future or should we remove it since we are changing the on-disk format here anyway?

Sorry, I have no idea why it was introduced in the first place. Dropping it now seems like a good idea. I don't see a future use-case as of now.

**Comment by CongMinYin** [comment]
> Additional concerns:
> 
> * `WriteLogCacheEntry::entry_index` is also on-disk and used by the `ssd` mode to store the log entry pointer but is still uint32_t
> * `entry_index` parameter of `AbstractWriteLog::update_entry()` is int but the `ssd` mode passes uint64_t
> 
> Please perform a thorough review of all fields, parameters and local variables across the entire `src/librbd/cache/pwl` code base to make sure there is nothing else of this sort lingering out there. Especially the on-disk fields -- I'd rather not change the on-disk format again...

I didn't find `entry_index` used by `ssd` mode. Can you point out where it is? In fact, `entry_index` doesn't work. We can delete it. It's just assigned, not used anywhere else :
https://github.com/ceph/ceph/blob/3111ba871f58996c1564c68d53f6c50cee922a01/src/librbd/cache/pwl/rwl/WriteLog.cc#L113
And the author Lisa also comment here that time, it can be remove.
https://github.com/ceph/ceph/blob/3111ba871f58996c1564c68d53f6c50cee922a01/src/librbd/cache/pwl/Types.h#L233

parameter `entry_index` in `update_entries()` should be change to `uint64_t`, done.

**Comment by CongMinYin** [comment]
add a new commit to remove `cur_sync_gen`

**Comment by idryomov** [comment]
> I didn't find `entry_index` used by `ssd` mode. Can you point out where it is?

Sorry, I think I got confused.  Having two different but identically named classes (`WriteLog`) in identically named files (`WriteLog.cc`) just in different directories is very misleading.

**Comment by idryomov** [comment]
But since the definition of `WriteLogPoolRoot` is shared and there is no explicit en/decoder for `rwl` mode, when we make the log entry pointers larger aren't we also changing the on-disk format for `rwl` mode?

**Comment by idryomov** [comment]
On top of that, aren't things like
```
  #ifdef WITH_RBD_RWL
  TOID(uint8_t) write_data;
  #endif
  #ifdef WITH_RBD_SSD_CACHE
  uint64_t write_data_pos = 0; /* SSD data offset */
  #endif
```
mean that simply enabling `WITH_RBD_SSD_CACHE` together with `WITH_RBD_RWL` already changes the on-disk format for `rwl` mode because additional fields get added?

**Comment by MahatiC** [comment]
> Having two different but identically named classes (`WriteLog`) in identically named files (`WriteLog.cc`) just in different directories is very misleading.

Hi Ilya, this design of having identical files was done in collaboration with upstream members at the time. In fact it was an explicit request. So I guess it's a bit subjective :)

**Comment by idryomov** [comment]
I'm not saying it's wrong -- just that it confused me when I was grepping for `entry_index`.

**Comment by MahatiC** [comment]
> But since the definition of `WriteLogPoolRoot` is shared and there is no explicit en/decoder for `rwl` mode, when we make the log entry pointers larger aren't we also changing the on-disk format for `rwl` mode?

correct, we are. I think this means `rwl` version should also bump up.

**Comment by idryomov** [comment]
What about the side effect of compiling with `WITH_RBD_SSD_CACHE=ON`?  The `rwl` version would stay the same but the user wouldn't be able to reopen it anyway.  Since the on-disk definitions are shared this way, does it even make sense to have two different versions?

Or should we split the on-disk definitions into something like `WriteLogPoolRootRWL`, `WriteLogPoolRootSSD`, etc?

**Comment by MahatiC** [comment]
> What about the side effect of compiling with `WITH_RBD_SSD_CACHE=ON`? The `rwl` version would stay the same but the user wouldn't be able to reopen the cache because the offsets would change.

why would the offsets change in ssd? We allocate 4k blocks. Currently max `log_entries` per control block is 32. We can accommodate that number to include the sizes of both cases.

> Or should we split the on-disk definitions into something like `WriteLogPoolRootRWL`, `WriteLogPoolRootSSD`, etc?

I suppose this would cause a lot of changes to the code. And I think this change would also make code sharing between `rwl` and `ssd` modes difficult

**Comment by idryomov** [comment]
I was taking about the `rwl` mode.  For example in `WriteLogCacheEntry`, the on-disk offset of `flags` in `rwl` mode depends on how Ceph is compiled: if `WITH_RBD_SSD_CACHE` is set, `write_data_pos` is added in front of `flags`.

In `ssd` mode the on-disk layout is defined by the en/decoder so it isn't sensitive to this `ifdef`ery.

**Comment by CongMinYin** [comment]
@idryomov @MahatiC ， After our discussion, it's not only the change of version that will lead to the offset problem of disk layout. Even if the version remains unchanged, for example, changing the compilation options from both enable `-DWITH_RBD_RWL=ON -DWITH_RBD_SSD_CACHE=ON` to only enable `-DWITH_RBD_RWL=ON`  will also cause this problem. For hard disk data, it is bad to use any compiler related methods or variables. But all persistent data structures used `#if_def` including `WriteLogPoolRoot` and `WriteLogCacheEntry`.  There are two ways:
1. use `union` to instead, and put `union` below common field which cann't be changed any more. (recommend)
2. separate `rwl` and `ssd` data structure, but it need modify lots of codes.

**Comment by CongMinYin** [comment]
@idryomov Anothing, the current cache code requires the same version, otherwise it is not available, see:
https://github.com/ceph/ceph/blob/3e2c8e94fb9fb8421a08ca425b14833a981565a6/src/librbd/cache/pwl/rwl/WriteLog.cc#L336
Because the later plan is to develop replica for RWL, the hard disk layout will be modified, that is https://github.com/ceph/ceph/pull/41432. Replica is what the community and customers want.

**Comment by CongMinYin** [comment]
@idryomov @MahatiC I tried to write a version of layout data structure to get rid of `#ifdef`.  Could you please take a look?
And @MahatiC Could I merge layout version to one field?

File: src/librbd/cache/pwl/Types.h
Comment (MahatiC) [inline src/librbd/cache/pwl/Types.h:234]: @CongMinYin Does this work if code is compiled only with `WITH_RBD_SSD_CACHE`?

File: src/librbd/cache/pwl/Types.h
Comment (MahatiC) [inline src/librbd/cache/pwl/Types.h:302]: what is the use-case for `unused` variable here?

File: src/librbd/cache/pwl/Types.h
Comment (MahatiC) [inline src/librbd/cache/pwl/Types.h:286]: this looks fine to me. AFAIK the `_u64` variable wasn't used in `rwl` mode

File: src/librbd/cache/pwl/Types.h
Comment (CongMinYin) [reply src/librbd/cache/pwl/Types.h:302]: > what is the use-case for `unused` variable here?

It is unused, but it won't take up any extra space because it is in `union`. I just want to keep the SSD structure so that we can conveniently extend the fields in the future.

**Comment by CongMinYin** [comment]
I changed some fields to 64 bits, which is not necessary now, but they are expected not to be changed in the future, so maybe we need to discuss here. For example, layout version, actually 8 bits  is enough.

File: src/librbd/cache/pwl/Types.h
Comment (CongMinYin) [reply src/librbd/cache/pwl/Types.h:234]: > @CongMinYin Does this work if code is compiled only with `WITH_RBD_SSD_CACHE`?

If change to union, it is independent of compilation options.

Comment (MahatiC) [reply src/librbd/cache/pwl/Types.h:234]: I just tried to compile this PR. This PR does not compile in `WITH_RBD_SSD_CACHE` mode. That was the reason for using the `ifdef` condition initially.

```
../src/librbd/cache/pwl/Types.h:233:7: error: 'PMEMoid' does not name a type
```

**Review by MahatiC — CHANGES_REQUESTED** [review]

File: src/librbd/cache/pwl/Types.h
Comment (CongMinYin) [reply src/librbd/cache/pwl/Types.h:234]: > ```
> ../src/librbd/cache/pwl/Types.h:233:7: error: 'PMEMoid' does not name a type
> ```
 Get your point, now I understand why use `#ifdef`. The available solution is to keep `#ifdef`, but leave it in the `union`, but remove `#ifdef` about `ssd`, because ssd use encode/decode, It will not be affected. So that, there are 2 case:
1) for rwl side, including `WITH_RBD_RWL`, union length is the bigger one. Whether the `WITH_RBD_SSD_CACHE` is compiled or not, union length is fixed. 
2) for ssd side, including `WITH_RBD_SSD_CACHE`,  `ssd` use encode/decode,  doesn't matter.

**Comment by CongMinYin** [comment]
@MahatiC Please take a look the updated code. I also compiled it only `WITH_RBD_SSD_CACHE`, no error.

**Comment by MahatiC** [comment]
> @MahatiC Please take a look the updated code. I also compiled it only `WITH_RBD_SSD_CACHE`, no error.

The recent change looks good and seems to address the versioning issue when both modes are enabled and also `ssd`/`rwl` combinations

**Comment by CongMinYin** [comment]
ping @idryomov

**Comment by CongMinYin** [comment]
@idryomov Rebased

I want to discuss with you about unit(`uint32_t`), I still suggest changing all the rest `uint32_t` to `uint_64`(mostly `data_len).` Today's memory is not expensive. We can waste a little for security and future scalability. On the other hand, I also thought of some possible problems due to `uint32_t`.  we can't guarantee that the value passed down from the upper layer is less than 4G, Extent `offset` and `length` is defined as `uint64_t`.  see:
https://github.com/ceph/ceph/blob/3e02117bade35b658f93aae905c98e69aa68148f/src/librbd/io/Types.h#L275
There are many variables using `uint32_t` like `data_len, block_size, alloc_size, discard_granularity_bytes, num_unpublished_reserves`. About write_same `data_len`, it is from `bl.length()`, There seems to be no guarantee that it will be less than 4G too. It has appeared many times. But I don't change it now.  

In the whole PWL, only `num_lanes, num_log_entries`, can be expressed as 32 bits, but about `entries`, in order to share fields with SSDs, it change to `uint64_t`, and other temporary variables changing to `uint64_t` will looks better.

File: src/librbd/cache/pwl/ssd/WriteLog.cc
Comment (ideepika) [inline src/librbd/cache/pwl/ssd/WriteLog.cc:245]: just trying to understand, why didn't we use `this->m_bytes_allocated_cap = this->m_log_pool_size -
        DATA_RING_BUFFER_OFFSET - MIN_WRITE_ALLOC_SSD_SIZE;` here like in previous case(ring buffer comment)?

Comment (ideepika) [reply src/librbd/cache/pwl/ssd/WriteLog.cc]: @CongMinYin

Comment (CongMinYin) [reply src/librbd/cache/pwl/ssd/WriteLog.cc]: yes, you are right. This is an error cause by automatic rebase. I didn't check it carefully. I meant to move the code, but git thinks it's a new code. Before rebased, other PR modified the code I want to move. I should check it more carefully

File: src/librbd/cache/pwl/Types.h
Comment (ideepika) [inline src/librbd/cache/pwl/Types.h:181]: this must be used for layout_version, which IIUC is either 0,1 or 2; should we keep same datatype then?

Comment (ideepika) [reply src/librbd/cache/pwl/Types.h:181]: @CongMinYin I meant using the smaller size data type: `uint8_t` since it can accommodate all conditions

Comment (CongMinYin) [reply src/librbd/cache/pwl/Types.h]: Modified. As a constant, 8 bytes is enough.

File: src/librbd/cache/pwl/Types.h
Comment (ideepika) [reply src/librbd/cache/pwl/Types.h:287]: @CongMinYin would be good to update here as well ` uint8_t layout_version;`

Comment (CongMinYin) [reply src/librbd/cache/pwl/Types.h:287]: Setting it to `uint8_t`  here does not save space. The previous code also used `uint64_t`. In a structure, there is a byte alignment rule. Because the next element is `uint64_t`, it will be aligned to an integer multiple address of 64 bits. Therefore, even if the version is `uint8_t`, it will occupy 64 bits of space. Because next  element will start from 64 bit aligned address. so why don't use `uint64_t`.

Comment (ideepika) [reply src/librbd/cache/pwl/Types.h:287]: I see, should be okay then, thanks for the explaination!

**Comment by ideepika** [comment]
https://pulpito.ceph.com/ideepika-2021-11-03_11:46:43-rbd-wip-ssd-cache-testing-distro-basic-smithi
https://pulpito.ceph.com/ideepika-2021-11-02_12:33:30-rbd-wip-ssd-cache-testing-distro-basic-smithi

cache failures:
https://tracker.ceph.com/issues/53110#change-205407
https://tracker.ceph.com/issues/49876

failures unrelated and tracked by:
https://tracker.ceph.com/issues/53108#change-205427
https://tracker.ceph.com/issues/51282#change-205397
https://tracker.ceph.com/issues/52503#change-205400
https://tracker.ceph.com/issues/53108#change-205427

File: src/librbd/cache/pwl/Types.h
Comment (trociny) [reply src/librbd/cache/pwl/Types.h:236]: I don't like this ` #ifdef WITH_RBD_RWL` in on-disk structure. I understand the reason and it seems not to cause the issue, as you explained. Still it looks at least very confusing.

And I am not sure it is a good idea to use "external" structure in on-disk structure. Are you sure `PMEMoid` will not change in the future? How are you going to track this?

Using encode/decode seems like the right solution indeed. Though I will not insist on changing this in this PR. I will not block the PR. Just wanted to bring this for discussion.

Comment (CongMinYin) [reply src/librbd/cache/pwl/Types.h:236]: > I don't like this ` #ifdef WITH_RBD_RWL` in on-disk structure. I understand the reason and it seems not to cause the issue, as you explained. Still it looks at least very confusing.
> 
> And I am not sure it is a good idea to use "external" structure in on-disk structure. Are you sure `PMEMoid` will not change in the future? How are you going to track this?
> 
> Using encode/decode seems like the right solution indeed. Though I will not insist on changing this in this PR. I will not block the PR. Just wanted to bring this for discussion.

I agree with you. This commit is quite old and not rigorous. As you pointed out, the external data type is unreliable and may be changed. Another side, we planned to change the libpmemobj library to libpmem to remove external types, use addresses and offsets, and do space management by ourselves, https://github.com/ceph/ceph/pull/41432.
@ideepika  So I want to remove the change of hard disk format,  keep version and memory format change. Hard disk format will be better solved in the future.

**Comment by CongMinYin** [comment]
@ideepika I remove the change disk format commit because it is not good enough and can be better solved in the future. Could you please retest?

**Comment by ideepika** [comment]
thanks @CongMinYin we also had concerns regarding them, I will do a retest

**Review by trociny — APPROVED** [review]
LGTM

**Comment by CongMinYin** [comment]
@idryomov As we discussed in tracker, I add  `qa/suites/rbd/persistent-writeback-cache/4-pool/cache-8GB.yaml`. + @trociny

**Comment by trociny** [comment]
> @idryomov As we discussed in tracker, I add `qa/suites/rbd/persistent-writeback-cache/4-pool/cache-8GB.yaml`. + @trociny

nit: I don't like very much 1/8GB in yaml file names. May be lets 1G test remain just `cache.yaml` (or become `small-cache.yaml`) and the 8GB `big-cache.yaml`? We already have examples like: `qemu/pool/small-cache-pool.yaml`, `basic/cachepool/small.yaml` ...

You may just ignore this comment if you like your variant more by some reason.

**Comment by CongMinYin** [comment]
> nit: I don't like very much 1/8GB in yaml file names. May be lets 1G test remain just `cache.yaml` (or become `small-cache.yaml`) and the 8GB `big-cache.yaml`? We already have examples like: `qemu/pool/small-cache-pool.yaml`, `basic/cachepool/small.yaml` ...

Good suggestion, done.

**Review by ideepika — APPROVED** [review]
thanks @CongMinYin !

**Comment by ideepika** [comment]
https://pulpito.ceph.com/ideepika-2021-11-12_16:20:59-rbd-wip-deepika-testing-2021-11-12-1203-distro-basic-smithi/
https://pulpito.ceph.com/ideepika-2021-11-12_08:56:59-rbd-wip-deepika-testing-2021-11-12-1203-distro-basic-smithi/

failures not related:
https://tracker.ceph.com/issues/53247
https://tracker.ceph.com/issues/53246
https://tracker.ceph.com/issues/52948

**Pull request merged**
